File: src/sw/sa_pipeline/stage7_sa.py
class Stage7SA:
    """
    Seventh Pipe from SA_Verilog. This pipe is responsible to execute the 2-2 
    additions for the distances found in the left pipe.
    """

    def __init__(self):
        """

        """
        self.new_output: dict = {
            'th_idx': 0,
            'th_valid': False,
            'dc': 0,
            'dvas': 0,
            'dvbs': 0
        }
        self.old_output: dict = self.new_output.copy()

    def compute(self, st6_input: dict):
        """

        @param st6_input:
        """
        # moving forward the ready outputs
        self.old_output = self.new_output.copy()

        st6_th_idx: int = st6_input['th_idx']
        st6_th_valid: bool = st6_input['th_valid']

        # fixme only for debugging
        if st6_th_idx == 0:
            z = 1

        st6_dvac: int = st6_input['dvac']
        st6_dvbc: int = st6_input['dvbc']
        st6_dvas: list = st6_input['dvas']
        st6_dvbs: list = st6_input['dvbs']

        dc: int = st6_dvac + st6_dvbc
        dvas: list = st6_dvas[0] + st6_dvas[1]
        dvbs: list = st6_dvbs[0] + st6_dvbs[1]

        self.new_output: dict = {
            'th_idx': st6_th_idx,
            'th_valid': st6_th_valid,
            'dc': dc,
            'dvas': dvas,
            'dvbs': dvbs
        }

File: src/cpp/sw/sa_pipeline/stage7_sa.cpp
#include <iostream>

class Stage7SA {
private:
    int* new_output;
    int* old_output;

public:
    Stage7SA() {
        new_output = new int[5]();
        old_output = new int[5]();
    }

    ~Stage7SA() {
        delete[] new_output;
        delete[] old_output;
    }

    void compute(int* st6_input) {
        std::copy(new_output, new_output + 5, old_output);

        int st6_th_idx = st6_input[0];
        bool st6_th_valid = st6_input[1];
        int st6_dvac = st6_input[2];
        int st6_dvbc = st6_input[3];
        int* st6_dvas = st6_input + 4;
        int* st6_dvbs = st6_input + 6;

        int dc = st6_dvac + st6_dvbc;
        int dvas = st6_dvas[0] + st6_dvas[1];
        int dvbs = st6_dvbs[0] + st6_dvbs[1];

        std::copy(st6_input, st6_input + 5, new_output);
        new_output[2] = dc;
        new_output[3] = dvas;
        new_output[4] = dvbs;
    }
};